lib/AST/ASTPrinter.cpp

```diff
@@ -4785,7 +4787,13 @@ void PrintAST::visitErrorExpr(ErrorExpr *expr) {
   Printer << "<error>";
 }
 
-void PrintAST::visitTernaryExpr(TernaryExpr *expr) {}
+void PrintAST::visitTernaryExpr(TernaryExpr *expr) {
+  if (expr->getCondExpr()) {
+    visit(expr->getThenExpr());
+  } else {
+    visit(expr->getElseExpr());
+  }
+}
```

Comment:
This doesn't seem correct to me; to clarify, it should be outputting the original source code for the expression node. So in this case you'll want to first print the `getCondExpr`, then print a `?`, then print the `getThenExpr`, then print a `:`, then print the `getElseExpr`.

Note though that there is some awkwardness to the representation of an unresolved TernaryExpr (prior to operator folding), the `getElseExpr` may be null. In that case, we should just be able to skip printing it, and it should be handled by `SequenceExpr` printing.

Reply:
> Note though that there is some awkwardness to the representation of an unresolved TernaryExpr (prior to operator folding), the `getCondExpr` and `getElseExpr` may be null. In that case, we should just be able to skip printing them, and it should be handled by `SequenceExpr` printing.

Could you elaborate on `SequenceExpr`?

Reply:
> Could you elaborate on `SequenceExpr`?

SequenceExpr is a node we produce when parsing, and is used to represent a binary operator chain such as `x + y + z` (producing `{Expr, Op, Expr, Op, Expr}`), but also includes ternaries, where we produce a sequence of `{CondExpr, TernaryExpr, ElseExpr}`. Before type-checking, we fold SequenceExpr into the correct structure (e.g creating `BinaryExpr` nodes), for ternaries this involves setting the CondExpr and ElseExpr on the ternary rather than having them be elements of the sequence. This is done outside the parser as looking up operator precedences is required to correctly fold, and that requires doing name lookups.

So the correct way of printing this would be to check to make sure `getCondExpr` and `getElseExpr` are non-null when printing them. And then for SequenceExpr, we'll just want to print each element in turn.

I should note that SequenceExpr only exists in un-type-checked AST, so it won't appear with `-print-ast`, unless there is potentially a semantic error that prevents it from being folded. I think we still ought to handle it for completeness sake though.

Reply:
Thanks for the explanation!